src/libsyntax/parse/parser.rs

```diff
@@ -5060,7 +5061,17 @@ impl<'a> Parser<'a> {
                 }
             }
         }
-        Ok((ident, ItemKind::Trait(unsafety, tps, bounds, trait_items), None))
+        Ok((ident, ItemKind::Trait(IsAuto::No, unsafety, tps, bounds, trait_items), None))
+    }
+
+    fn parse_item_auto_trait(&mut self, unsafety: Unsafety) -> PResult<'a, ItemInfo> {
```

Comment:
`auto` traits should be parsed exactly like other traits, if they need extra restrictions (e.g. no generics), they can be added in `librustc_passes/ast_validation.rs`.
This way we can get proper error messages instead of generic `expected x, found y` from the parser.